File: src/main/java/pw/switchcraft/plethora/gameplay/modules/glasses/objects/object2d/ObjectGroup2d.java
package pw.switchcraft.plethora.gameplay.modules.glasses.objects.object2d;

import com.google.common.base.Objects;
import it.unimi.dsi.fastutil.ints.IntSet;
import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.minecraft.client.util.math.MatrixStack;
import net.minecraft.network.PacketByteBuf;
import pw.switchcraft.plethora.gameplay.modules.glasses.canvas.CanvasClient;
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.BaseObject;
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.ObjectGroup;
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.ObjectRegistry;
import pw.switchcraft.plethora.util.ByteBufUtils;
import pw.switchcraft.plethora.util.Vec2d;

import javax.annotation.Nonnull;

public class ObjectGroup2d extends BaseObject implements ObjectGroup.Group2d, Positionable2d {
	private Vec2d position = Vec2d.ZERO;

	public ObjectGroup2d(int id, int parent) {
		super(id, parent, ObjectRegistry.GROUP_2D);
	}

	@Nonnull
	@Override
	public Vec2d getPosition() {
		return position;
	}

	@Override
	public void setPosition(@Nonnull Vec2d position) {
		if (!Objects.equal(this.position, position)) {
			this.position = position;
			setDirty();
		}
	}

	@Override
	public void writeInitial(PacketByteBuf buf) {
		ByteBufUtils.writeVec2d(buf, position);
	}

	@Override
	public void readInitial(PacketByteBuf buf) {
		position = ByteBufUtils.readVec2d(buf);
	}

	@Override
	@Environment(EnvType.CLIENT)
	public void draw(CanvasClient canvas, MatrixStack matrices) {
		IntSet children = canvas.getChildren(id());
		if (children == null) return;

		matrices.push();
		matrices.translate(position.x(), position.y(), 0);

		canvas.drawChildren(children.iterator(), matrices);

		matrices.pop();
	}
}

File: src/main/java/pw/switchcraft/plethora/gameplay/modules/glasses/objects/object2d/ObjectGroup2d.kt
package pw.switchcraft.plethora.gameplay.modules.glasses.objects.object2d

import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.minecraft.client.render.VertexConsumerProvider
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.network.PacketByteBuf
import pw.switchcraft.plethora.gameplay.modules.glasses.canvas.CanvasClient
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.BaseObject
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.ObjectGroup.Group2d
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.ObjectRegistry.GROUP_2D
import pw.switchcraft.plethora.util.ByteBufUtils
import pw.switchcraft.plethora.util.DirtyingProperty
import pw.switchcraft.plethora.util.Vec2d
import javax.annotation.Nonnull

class ObjectGroup2d(
  id: Int,
  parent: Int
) : BaseObject(id, parent, GROUP_2D), Group2d, Positionable2d {
  override var position by DirtyingProperty(Vec2d.ZERO!!)

  override fun writeInitial(@Nonnull buf: PacketByteBuf) {
    ByteBufUtils.writeVec2d(buf, position)
  }

  override fun readInitial(@Nonnull buf: PacketByteBuf) {
    position = ByteBufUtils.readVec2d(buf)
  }

  @Environment(EnvType.CLIENT)
  override fun draw(canvas: CanvasClient, matrices: MatrixStack, consumers: VertexConsumerProvider?) {
    val children = canvas.getChildren(id) ?: return
    matrices.push()
    matrices.translate(position.x(), position.y(), 0.0)
    canvas.drawChildren(children.iterator(), matrices, consumers)
    matrices.pop()
  }
}
